stellargraph/core/graph.py

```diff
@@ -904,7 +904,34 @@ def _adjacency_types(self, graph_schema: GraphSchema):
         if self._graph is not None:
             return self._graph.adjacency_types(graph_schema)
 
-        raise NotImplementedError()
+        source_types, rel_types, target_types = self._edge_type_triples(slice(None))
+
+        triples = defaultdict(lambda: defaultdict(lambda: []))
+
+        iterator = zip(
+            source_types,
+            rel_types,
+            target_types,
+            self._edges.sources,
+            self._edges.targets,
+        )
+        for src_type, rel_type, tgt_type, src, tgt in iterator:
+            triple = EdgeType(src_type, rel_type, tgt_type)
+            triple_dict = triples[triple]
+            triple_dict[src].append(tgt)
+
+            if not self.is_directed() and src != tgt:
+                # non-self loop in an undirected graph, so we should add the other direction too
+                triple_dict[tgt].append(src)
+
+                other_triple = EdgeType(tgt_type, rel_type, src_type)
+                if other_triple != triple:
+                    # the edge type differs too, so we need to add its other direction
+                    other_triple_dict = triples[other_triple]
+                    other_triple_dict[src].append(tgt)
+                    other_triple_dict[tgt].append(src)
```

Comment:
This means the result of grabbing `triples[(src_type, rel_type, tgt_type)][node]` can contain either `src_type` or `tgt_type` nodes rather than nodes of `tgt_type` only, which to me seems less useful for things like HinSAGE where you want to ensure doing a neighbourhood hop only returns nodes of one type. I'd consider `EdgeType(src_type, rel_type, tgt_type)`as inherently having some sort of directedness even in undirected graphs, did you consider just doing:
```python
if not self.is_directed() and src != tgt:
    other_triple = EdgeType(tgt_type, rel_type, src_type)
    if other_triple != triple:
        other_triple_dict = triples[other_triple]
        other_triple_dict[tgt].append(src)
```
I haven't looked closely at what the existing networkx based implementation is doing, but I would've thought it would be something along those lines

Reply:
Ah, yeah, you're right, I must've not actually checked against the implementation correctly...